File: day11_part2_2020.dart

import 'dart:io';

void main() {
  List<String> seating = File('input.txt').readAsLinesSync();

  List<List<String>> grid = seating.map((row) => row.split('')).toList();

  int rows = grid.length;
  int cols = grid[0].length;

  List<List<String>> newGrid = List<List<String>>.generate(rows, (index) => List<String>.filled(cols, '.'));

  bool changed = true;

  while (changed) {
    changed = false;

    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        int occupiedSeats = 0;

        for (int x = -1; x <= 1; x++) {
          for (int y = -1; y <= 1; y++) {
            if (x == 0 && y == 0) continue;

            int dx = i + x;
            int dy = j + y;

            while (dx >= 0 && dx < rows && dy >= 0 && dy < cols) {
              if (grid[dx][dy] == '#') {
                occupiedSeats++;
                break;
              } else if (grid[dx][dy] == 'L') {
                break;
              }

              dx += x;
              dy += y;
            }
          }
        }

        if (grid[i][j] == 'L' && occupiedSeats == 0) {
          newGrid[i][j] = '#';
          changed = true;
        } else if (grid[i][j] == '#' && occupiedSeats >= 5) {
          newGrid[i][j] = 'L';
          changed = true;
        } else {
          newGrid[i][j] = grid[i][j];
        }
      }
    }

    for (int i = 0; i < rows; i++) {
      for (int j = 0; j < cols; j++) {
        grid[i][j] = newGrid[i][j];
      }
    }
  }

  int occupiedSeats = 0;
  for (int i = 0; i < rows; i++) {
    for (int j = 0; j < cols; j++) {
      if (grid[i][j] == '#') {
        occupiedSeats++;
      }
    }
  }

  print(occupiedSeats);
}

File: objectivec/day11_part2_2020.m

#import <Foundation/Foundation.h>

static const int dirs[8][2] = {{-1,-1},{0,-1},{1,-1},{-1,0},{1,0},{-1,1},{0,1},{1,1}};

int countVisible(NSArray<NSMutableString *> *grid, int r, int c) {
    int rows = (int)grid.count, cols = (int)[grid[0] length], cnt = 0;
    for (int d = 0; d < 8; ++d) {
        int dr = dirs[d][1], dc = dirs[d][0];
        int rr = r + dr, cc = c + dc;
        while (rr >= 0 && rr < rows && cc >= 0 && cc < cols) {
            unichar ch = [grid[rr] characterAtIndex:cc];
            if (ch == 'L') break;
            if (ch == '#') { ++cnt; break; }
            rr += dr; cc += dc;
        }
    }
    return cnt;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSError *err = nil;
        NSString *txt = [NSString stringWithContentsOfFile:@"input.txt"
                                                  encoding:NSUTF8StringEncoding
                                                     error:&err];
        if (!txt) { NSLog(@"%@", err); return 1; }
        NSMutableArray<NSMutableString *> *grid = [NSMutableArray array];
        for (NSString *line in [txt componentsSeparatedByCharactersInSet:[NSCharacterSet newlineCharacterSet]]) {
            if (line.length) [grid addObject:[line mutableCopy]];
        }

        BOOL changed = YES;
        while (changed) {
            changed = NO;
            NSMutableArray<NSMutableString *> *next = [NSMutableArray arrayWithCapacity:grid.count];
            for (NSMutableString *row in grid) [next addObject:[row mutableCopy]];
            int rows = (int)grid.count, cols = (int)[grid[0] length];
            for (int i = 0; i < rows; ++i) {
                for (int j = 0; j < cols; ++j) {
                    unichar seat = [grid[i] characterAtIndex:j];
                    if (seat == 'L' && countVisible(grid, i, j) == 0) {
                        [next[i] replaceCharactersInRange:NSMakeRange(j, 1) withString:@"#"];
                        changed = YES;
                    } else if (seat == '#' && countVisible(grid, i, j) >= 5) {
                        [next[i] replaceCharactersInRange:NSMakeRange(j, 1) withString:@"L"];
                        changed = YES;
                    }
                }
            }
            grid = next;
        }

        int occupied = 0;
        for (NSMutableString *row in grid) {
            for (NSUInteger i = 0; i < row.length; ++i)
                if ([row characterAtIndex:i] == '#') ++occupied;
        }
        printf("%d\n", occupied);
    }
    return 0;
}
